File: src/lib/verification/arcane/evidence-envelope.mjs
// Evidence envelope — seals evidence-capability receipts and carries the
// full Arcane-side dependency record the frozen receipt cannot express.
//
// Arcane tracks dependency edges across a much wider set of dimensions than
// the frozen `evidence-capability-receipt-v1` schema's
// `dependsOn[].kind` enum allows (six values only: decision | source-revision
// | config-digest | tool-digest | policy-digest | evidence). Per
// INTERFACES.md's frozen-contract constraint, this module never edits that
// schema. Instead:
//
//   - the full dimension set lives here as DEPENDENCY_DIMENSION and is what
//     `invalidation.mjs`'s DependencyLedger actually indexes and cascades on;
//   - `sealEvidence` projects each dependency down to one of the six legal
//     `kind` values via the explicit DEPENDENCY_KIND_PROJECTION table below,
//     and folds the original dimension into the receipt's free-form `ref`
//     string (`"<dimension>:<ref>"`) so no information is silently discarded
//     even where the *kind* bucket is lossy;
//   - the lossless dependency list (dimension + ref + digest, untouched)
//     lives on the `envelope` returned alongside the receipt;
//   - every dimension whose projection is not faithful is flagged in the
//     table (`faithful: false`) — see the proposed contract amendment in the
//     S05 lane report, and PROPOSED_CONTRACT_AMENDMENT below for the same
//     text in code.

import { mintId as mintKernelId } from '../../core/kernel-binding.mjs';
import { assertValid, loadSchema } from '../../contracts/arcane/validate.mjs';
import { digestValue } from '../../contracts/arcane/canonical.mjs';
import { ArcaneError } from '../../contracts/arcane/errors.mjs';

// ---------------------------------------------------------------------------
// Dependency dimensions (canonical list; see INTERFACES.md evidence envelope)
// ---------------------------------------------------------------------------

/**
 * The full dependency-dimension set Arcane tracks for invalidation. This is
 * deliberately richer than the frozen receipt's `dependsOn[].kind` enum —
 * see the projection table below. 'evidence' is the transitive-cascade edge
 * kind: an evidence record that depends on another evidence record via this
 * dimension is invalidated whenever that other record goes stale, regardless
 * of digest (DependencyLedger cascades structurally through these edges).
 */
export const DEPENDENCY_DIMENSION = Object.freeze([
  'source-digest', // source/file/object digest
  'git-revision', // git revision
  'worktree-dirty', // worktree dirty state
  'lockfile-version', // lockfile/dependency version
  'config-digest', // config
  'schema-version', // schema (of the artifact/contract under evidence, not this package's own schemas)
  'fixture-digest', // fixture/test data
  'tool-version', // tool/provider/runtime version
  'environment', // environment/host/platform
  'method-fingerprint', // method fingerprint
  'policy-version', // policy/catalog/schema version
  'topology', // target/component topology
  'external-source-revision', // external source revision
  'approval-authority', // approval/waiver authority
  'upstream-contract-revision', // upstream contract revision
  'evidence', // transitive edge onto another evidence record
]);

const DIMENSION_SET = new Set(DEPENDENCY_DIMENSION);

// ---------------------------------------------------------------------------
// Projection table: full dimension -> frozen receipt `dependsOn[].kind`
// ---------------------------------------------------------------------------

// Derived from the frozen schema itself (never hardcoded) so this table can
// never silently drift from what evidence-capability-receipt-v1.schema.json
// actually allows.
const FROZEN_KIND = Object.freeze(
  loadSchema('evidence-capability-receipt-v1').properties.dependsOn.items.properties.kind.enum,
);

function kind(value) {
  if (!FROZEN_KIND.includes(value)) {
    throw new ArcaneError('ARC_SCHEMA_INVALID', `projection target not in frozen dependsOn[].kind enum: ${value}`, {
      value,
      allowed: FROZEN_KIND,
    });
  }
  return value;
}

/**
 * dimension -> { kind, faithful, note }. `faithful: false` means the six-value
 * enum has no bucket that actually means this dimension — the projection is a
 * best-fit placement so the fact is not lost, not a claim of semantic
 * equivalence. See PROPOSED_CONTRACT_AMENDMENT.
 */
export const DEPENDENCY_KIND_PROJECTION = Object.freeze({
  'source-digest': { kind: kind('source-revision'), faithful: true, note: 'a content digest identifying source state is what source-revision means' },
  'git-revision': { kind: kind('source-revision'), faithful: true, note: 'exact match' },
  'worktree-dirty': { kind: kind('source-revision'), faithful: false, note: 'a dirty/clean flag is not a revision identifier; no faithful bucket exists' },
  'lockfile-version': { kind: kind('config-digest'), faithful: false, note: 'dependency-graph version is a distinct concept from generic config' },
  'config-digest': { kind: kind('config-digest'), faithful: true, note: 'exact match' },
  'schema-version': { kind: kind('config-digest'), faithful: false, note: 'a contract/schema version is not config' },
  'fixture-digest': { kind: kind('source-revision'), faithful: false, note: 'test/fixture data digest conflated with source-code revision' },
  'tool-version': { kind: kind('tool-digest'), faithful: true, note: 'version stands in for digest; close enough' },
  'environment': { kind: kind('tool-digest'), faithful: false, note: 'host/platform identity is not a tool' },
  'method-fingerprint': { kind: kind('tool-digest'), faithful: false, note: 'describes the producer\'s method, not tool identity' },
  'policy-version': { kind: kind('policy-digest'), faithful: true, note: 'version stands in for digest; close enough' },
  'topology': { kind: kind('config-digest'), faithful: false, note: 'structural target/component topology is not a config value' },
  'external-source-revision': { kind: kind('source-revision'), faithful: true, note: 'exact match' },
  'approval-authority': { kind: kind('decision'), faithful: false, note: 'an approval/waiver authority binding is not itself a sealed decision' },
  'upstream-contract-revision': { kind: kind('decision'), faithful: false, note: 'a contract revision is closer to source-revision than decision, but contracts derive from decisions; ambiguous either way' },
  evidence: { kind: kind('evidence'), faithful: true, note: 'exact match; the only dimension the frozen enum models precisely for transitive edges' },
});

// Dimensions with no faithful projection — the exact list the lane report's
// proposed contract amendment must justify extending the enum for.
export const UNFAITHFUL_DIMENSIONS = Object.freeze(
  Object.entries(DEPENDENCY_KIND_PROJECTION)
    .filter(([, v]) => !v.faithful)
    .map(([d]) => d),
);

/**
 * Proposed contract amendment text (also reproduced in the lane report).
 * Not applied anywhere — packages/contracts/ is read-only to this lane.
 */
export const PROPOSED_CONTRACT_AMENDMENT =
  'Extend evidence-capability-receipt-v1.dependsOn[].kind with a 7th value, ' +
  '"dimension-digest", paired with a new required-when-kind-is-dimension-digest ' +
  'sibling field `dimension` drawn from the full DEPENDENCY_DIMENSION set ' +
  '(packages/arcane/lib/evidence-envelope.mjs). This avoids re-litigating the ' +
  'enum every time a new invalidation dimension is discovered (9 of 16 current ' +
  'dimensions already have no faithful bucket: ' +
  Object.keys(DEPENDENCY_KIND_PROJECTION).filter((d) => !DEPENDENCY_KIND_PROJECTION[d].faithful).join(', ') +
  '), while leaving the existing 6 literal kinds untouched for callers that do ' +
  'not need dimension-level precision.';

function projectDependency({ dimension, ref, digest }) {
  if (!DIMENSION_SET.has(dimension)) {
    throw new ArcaneError('ARC_DEPENDENCY_UNKNOWN', `unknown dependency dimension: ${dimension}`, { dimension });
  }
  if (typeof ref !== 'string' || ref.length === 0) {
    throw new ArcaneError('ARC_SCHEMA_INVALID', 'dependency ref must be a non-empty string', { dimension, ref });
  }
  const projection = DEPENDENCY_KIND_PROJECTION[dimension];
  // Fold the original dimension into `ref` so the projection never silently
  // discards which dimension actually changed, even though `kind` collapses
  // several dimensions into the same bucket.
  return { kind: projection.kind, ref: `${dimension}:${ref}` };
}

// ---------------------------------------------------------------------------
// sealEvidence / envelopeDigest (INTERFACES.md §S05, binding signatures)
// ---------------------------------------------------------------------------

/**
 * Seal an observation into an evidence-capability receipt plus the richer
 * Arcane-side envelope.
 * @returns {{receipt: object, envelope: object}}
 */
export function sealEvidence({
  runId,
  taskId = null,
  contractId = null,
  producerAuthority,
  capability,
  observation,
  evidenceClass,
  sourceRevision,
  dependencies = [],
  authentication,
  replayDefense,
  observedAt,
}) {
  const { id: evidenceId, provisional } = mintKernelId('evidenceReceipt');
  const dependsOn = dependencies.map(projectDependency);

  const receipt = {
    schemaVersion: 1,
    kind: 'legion-evidence-capability-receipt',
    evidenceId,
    runId,
    taskId,
    contractId,
    producerAuthority,
    capability,
    observation,
    evidenceClass,
    authentication,
    replayDefense,
    sourceRevision,
    dependsOn,
    stale: false,
    observedAt,
  };
  assertValid('evidence-capability-receipt-v1', receipt);

  const envelopeCore = {
    evidenceId,
    runId,
    taskId,
    contractId,
    producerAuthority,
    capability,
    observation,
    evidenceClass,
    sourceRevision,
    // Lossless dependency list — full dimension set, untouched digests.
    dependencies: dependencies.map((d) => ({ dimension: d.dimension, ref: d.ref, digest: d.digest })),
    authentication,
    replayDefense,
    observedAt,
    stale: false,
    provisionalId: provisional,
    trustClass: authentication?.verificationMethod === 'unauthenticated' ? 'unauthenticated' : 'authenticated',
    receiptDigest: digestValue(receipt),
  };

  const envelope = { ...envelopeCore, envelopeDigest: envelopeDigest(envelopeCore) };
  return { receipt, envelope };
}

/** Digest of an envelope's content (excludes any pre-existing envelopeDigest field, so it is idempotent to call on an already-digested envelope). */
export function envelopeDigest(envelope) {
  const { envelopeDigest: _drop, ...rest } = envelope;
  return digestValue(rest);
}

// ---------------------------------------------------------------------------
// Legacy import (WP4 action 7) — additive helper, not in the binding list.
// ---------------------------------------------------------------------------

/**
 * Import a legacy-envelope-v1-shaped record as an Arcane evidence envelope
 * that can never gain trust. S00 finding: every legacy predecessor
 * `signature_or_mac` is a self-hash with no secret key — a legacy envelope's
 * `provenance.authenticated` is structurally `false` (the frozen
 * legacy-envelope-v1 schema enforces this with `const: false`). Migration
 * never upgrades that; it is not re-emitted as a
 * `evidence-capability-receipt-v1` claiming real authentication (that would
 * be the exact trust-laundering S00 warns against). It becomes an
 * envelope-only record, `trustClass: 'legacy-unauthenticated'`,
 * `neverQualifying: true`, importable into DependencyLedger only with
 * `{ trusted: false }` so it can be tracked/cascaded against but can never
 * make a criterion `proven`.
 */
export function importLegacyEvidence(legacyEnvelope, { runId = null, capability = 'legacy-import', observedAt } = {}) {
  if (!legacyEnvelope || legacyEnvelope.kind !== 'legion-legacy-envelope') {
    throw new ArcaneError('ARC_SCHEMA_INVALID', 'expected a legacy-envelope-v1 shaped record (kind: legion-legacy-envelope)', {
      kind: legacyEnvelope?.kind ?? null,
    });
  }
  if (legacyEnvelope.provenance?.authenticated !== false) {
    throw new ArcaneError('ARC_AUTH_LEGACY_DIGEST', 'legacy envelope provenance.authenticated must be false', {
      authenticated: legacyEnvelope.provenance?.authenticated ?? null,
    });
  }
  const { id: evidenceId } = mintKernelId('evidenceReceipt');
  return Object.freeze({
    evidenceId,
    runId,
    legacyKind: legacyEnvelope.legacyKind,
    capability,
    observation: Object.freeze({ legacyPayloadDigest: digestValue(legacyEnvelope.payload) }),
    evidenceClass: 'external',
    sourceRevision: legacyEnvelope.provenance.sourceRevision ?? null,
    dependencies: [],
    authentication: Object.freeze({
      issuerIdentity: legacyEnvelope.provenance.source,
      verificationMethod: 'unauthenticated',
      perMessage: false,
      verifiedAt: null,
    }),
    trustClass: 'legacy-unauthenticated',
    neverQualifying: true,
    provisionalMappingRef: legacyEnvelope.provisionalMappingRef ?? null,
    observedAt: observedAt ?? legacyEnvelope.provenance.capturedAt,
    stale: false,
  });
}

File: engine/crates/legion-policy/src/wf_port/wf072/evidence_envelope.rs
//! Faithful port of `src/lib/verification/arcane/evidence-envelope.mjs`.
//!
//! Seals evidence-capability receipts and carries the full Arcane-side
//! dependency record the frozen receipt cannot express (16 dependency
//! dimensions, projected down to the frozen 6-value `dependsOn[].kind` enum).
//!
//! Deviations from the JS source (both necessary, both noted at point of
//! use):
//!   - `sealEvidence` calls `assertValid('evidence-capability-receipt-v1',
//!     receipt)` against the JSON-schema loaded from
//!     `packages/contracts/schemas/`. wf072 does not own a JSON-schema
//!     validator or that schema file, so `seal_evidence` here performs the
//!     receipt's *structural* shape checks it can faithfully reproduce
//!     (required fields present) without re-implementing full JSON-schema
//!     validation. This is a scope gap, not a behavior change — see the
//!     wf072 report.
//!   - `mintKernelId` (Kernel-binding-based, provisional-when-unbound id
//!     minting) is out of this chunk's ownership; `seal_evidence` takes the
//!     minted `evidence_id` and `provisional` flag as caller-supplied
//!     inputs instead of minting them itself. This keeps the module pure
//!     and deterministic for testing, and the caller (once wired to a real
//!     Kernel binding) supplies exactly what `mintKernelId('evidenceReceipt')`
//!     would have produced.

use super::support::{digest_value, ArcCode, Json};

/// The full dependency-dimension set Arcane tracks for invalidation.
/// Mirrors JS `DEPENDENCY_DIMENSION` exactly (order and membership).
pub const DEPENDENCY_DIMENSION: &[&str] = &[
    "source-digest",
    "git-revision",
    "worktree-dirty",
    "lockfile-version",
    "config-digest",
    "schema-version",
    "fixture-digest",
    "tool-version",
    "environment",
    "method-fingerprint",
    "policy-version",
    "topology",
    "external-source-revision",
    "approval-authority",
    "upstream-contract-revision",
    "evidence",
];

/// Frozen receipt `dependsOn[].kind` enum (6 values), mirroring
/// `evidence-capability-receipt-v1.schema.json`'s `dependsOn.items.kind.enum`.
pub const FROZEN_KIND: &[&str] =
    &["decision", "source-revision", "config-digest", "tool-digest", "policy-digest", "evidence"];

/// dimension -> (kind, faithful, note). Mirrors JS `DEPENDENCY_KIND_PROJECTION`.
pub fn dependency_kind_projection(dimension: &str) -> Option<(&'static str, bool, &'static str)> {
    Some(match dimension {
        "source-digest" => ("source-revision", true, "a content digest identifying source state is what source-revision means"),
        "git-revision" => ("source-revision", true, "exact match"),
        "worktree-dirty" => ("source-revision", false, "a dirty/clean flag is not a revision identifier; no faithful bucket exists"),
        "lockfile-version" => ("config-digest", false, "dependency-graph version is a distinct concept from generic config"),
        "config-digest" => ("config-digest", true, "exact match"),
        "schema-version" => ("config-digest", false, "a contract/schema version is not config"),
        "fixture-digest" => ("source-revision", false, "test/fixture data digest conflated with source-code revision"),
        "tool-version" => ("tool-digest", true, "version stands in for digest; close enough"),
        "environment" => ("tool-digest", false, "host/platform identity is not a tool"),
        "method-fingerprint" => ("tool-digest", false, "describes the producer's method, not tool identity"),
        "policy-version" => ("policy-digest", true, "version stands in for digest; close enough"),
        "topology" => ("config-digest", false, "structural target/component topology is not a config value"),
        "external-source-revision" => ("source-revision", true, "exact match"),
        "approval-authority" => ("decision", false, "an approval/waiver authority binding is not itself a sealed decision"),
        "upstream-contract-revision" => ("decision", false, "a contract revision is closer to source-revision than decision, but contracts derive from decisions; ambiguous either way"),
        "evidence" => ("evidence", true, "exact match; the only dimension the frozen enum models precisely for transitive edges"),
        _ => return None,
    })
}

/// Dimensions with no faithful projection. Mirrors JS `UNFAITHFUL_DIMENSIONS`.
pub fn unfaithful_dimensions() -> Vec<&'static str> {
    DEPENDENCY_DIMENSION
        .iter()
        .filter(|d| matches!(dependency_kind_projection(d), Some((_, false, _))))
        .copied()
        .collect()
}

#[derive(Debug, Clone, PartialEq)]
pub struct Dependency {
    pub dimension: String,
    pub reference: String,
    pub digest: Option<String>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ProjectedDependency {
    pub kind: &'static str,
    pub reference: String,
}

#[derive(Debug, Clone, PartialEq)]
pub struct EnvelopeError {
    pub code: ArcCode,
    pub message: String,
}

/// Mirrors JS `projectDependency`.
fn project_dependency(dep: &Dependency) -> Result<ProjectedDependency, EnvelopeError> {
    let Some((kind, _faithful, _note)) = dependency_kind_projection(&dep.dimension) else {
        return Err(EnvelopeError {
            code: ArcCode::ArcDependencyUnknown,
            message: format!("unknown dependency dimension: {}", dep.dimension),
        });
    };
    if dep.reference.is_empty() {
        return Err(EnvelopeError {
            code: ArcCode::ArcSchemaInvalid,
            message: "dependency ref must be a non-empty string".to_string(),
        });
    }
    // Fold the original dimension into `ref` so the projection never
    // silently discards which dimension actually changed.
    Ok(ProjectedDependency { kind, reference: format!("{}:{}", dep.dimension, dep.reference) })
}

#[derive(Debug, Clone, PartialEq)]
pub struct SealEvidenceInput {
    pub evidence_id: String,
    pub provisional: bool,
    pub run_id: String,
    pub task_id: Option<String>,
    pub contract_id: Option<String>,
    pub producer_authority: String,
    pub capability: String,
    pub observation: Json,
    pub evidence_class: String,
    pub source_revision: Option<String>,
    pub dependencies: Vec<Dependency>,
    /// `authentication.verificationMethod`, if any — used only to derive
    /// `trustClass`, matching JS `authentication?.verificationMethod === 'unauthenticated'`.
    pub authentication_verification_method: Option<String>,
    pub authentication: Json,
    pub replay_defense: Json,
    pub observed_at: String,
}

#[derive(Debug, Clone, PartialEq)]
pub struct SealedEvidence {
    pub receipt: Json,
    pub envelope: Json,
    pub envelope_digest: String,
}

/// Seal an observation into an evidence-capability receipt plus the richer
/// Arcane-side envelope. Mirrors JS `sealEvidence`.
pub fn seal_evidence(input: SealEvidenceInput) -> Result<SealedEvidence, EnvelopeError> {
    let mut depends_on = Vec::with_capacity(input.dependencies.len());
    for dep in &input.dependencies {
        depends_on.push(project_dependency(dep)?);
    }

    let depends_on_json = Json::Arr(
        depends_on
            .iter()
            .map(|d| Json::Obj(vec![("kind".into(), Json::str(d.kind)), ("ref".into(), Json::str(d.reference.clone()))]))
            .collect(),
    );

    let receipt = Json::Obj(vec![
        ("schemaVersion".into(), Json::I64(1)),
        ("kind".into(), Json::str("legion-evidence-capability-receipt")),
        ("evidenceId".into(), Json::str(input.evidence_id.clone())),
        ("runId".into(), Json::str(input.run_id.clone())),
        ("taskId".into(), input.task_id.clone().map(Json::Str).unwrap_or(Json::Null)),
        ("contractId".into(), input.contract_id.clone().map(Json::Str).unwrap_or(Json::Null)),
        ("producerAuthority".into(), Json::str(input.producer_authority.clone())),
        ("capability".into(), Json::str(input.capability.clone())),
        ("observation".into(), input.observation.clone()),
        ("evidenceClass".into(), Json::str(input.evidence_class.clone())),
        ("authentication".into(), input.authentication.clone()),
        ("replayDefense".into(), input.replay_defense.clone()),
        ("sourceRevision".into(), input.source_revision.clone().map(Json::Str).unwrap_or(Json::Null)),
        ("dependsOn".into(), depends_on_json),
        ("stale".into(), Json::Bool(false)),
        ("observedAt".into(), Json::str(input.observed_at.clone())),
    ]);

    // Scope-limited structural check in place of full JSON-schema validation
    // (see module doc). runId/producerAuthority/capability/evidenceClass/
    // observedAt are the receipt's own non-nullable required fields.
    if input.run_id.is_empty()
        || input.producer_authority.is_empty()
        || input.capability.is_empty()
        || input.evidence_class.is_empty()
        || input.observed_at.is_empty()
    {
        return Err(EnvelopeError {
            code: ArcCode::ArcSchemaInvalid,
            message: "evidence-capability-receipt-v1 is missing a required field".to_string(),
        });
    }

    let receipt_digest = digest_value(&receipt);

    let trust_class = if input.authentication_verification_method.as_deref() == Some("unauthenticated") {
        "unauthenticated"
    } else {
        "authenticated"
    };

    let dependencies_json = Json::Arr(
        input
            .dependencies
            .iter()
            .map(|d| {
                Json::Obj(vec![
                    ("dimension".into(), Json::str(d.dimension.clone())),
                    ("ref".into(), Json::str(d.reference.clone())),
                    ("digest".into(), d.digest.clone().map(Json::Str).unwrap_or(Json::Null)),
                ])
            })
            .collect(),
    );

    let envelope_core = Json::Obj(vec![
        ("evidenceId".into(), Json::str(input.evidence_id.clone())),
        ("runId".into(), Json::str(input.run_id.clone())),
        ("taskId".into(), input.task_id.clone().map(Json::Str).unwrap_or(Json::Null)),
        ("contractId".into(), input.contract_id.clone().map(Json::Str).unwrap_or(Json::Null)),
        ("producerAuthority".into(), Json::str(input.producer_authority.clone())),
        ("capability".into(), Json::str(input.capability.clone())),
        ("observation".into(), input.observation.clone()),
        ("evidenceClass".into(), Json::str(input.evidence_class.clone())),
        ("sourceRevision".into(), input.source_revision.clone().map(Json::Str).unwrap_or(Json::Null)),
        ("dependencies".into(), dependencies_json),
        ("authentication".into(), input.authentication.clone()),
        ("replayDefense".into(), input.replay_defense.clone()),
        ("observedAt".into(), Json::str(input.observed_at.clone())),
        ("stale".into(), Json::Bool(false)),
        ("provisionalId".into(), Json::Bool(input.provisional)),
        ("trustClass".into(), Json::str(trust_class)),
        ("receiptDigest".into(), Json::str(receipt_digest)),
    ]);

    let env_digest = envelope_digest(&envelope_core);
    let envelope = with_envelope_digest(&envelope_core, &env_digest);

    Ok(SealedEvidence { receipt, envelope, envelope_digest: env_digest })
}

/// Digest of an envelope's content (excludes any pre-existing
/// `envelopeDigest` field, so it is idempotent to call on an
/// already-digested envelope). Mirrors JS `envelopeDigest`.
pub fn envelope_digest(envelope: &Json) -> String {
    let Json::Obj(pairs) = envelope else {
        return digest_value(envelope);
    };
    let rest: Vec<(String, Json)> = pairs.iter().filter(|(k, _)| k != "envelopeDigest").cloned().collect();
    digest_value(&Json::Obj(rest))
}

fn with_envelope_digest(core: &Json, env_digest: &str) -> Json {
    let Json::Obj(pairs) = core else { return core.clone() };
    let mut out = pairs.clone();
    out.push(("envelopeDigest".into(), Json::str(env_digest)));
    Json::Obj(out)
}

// ---------------------------------------------------------------------
// Legacy import (WP4 action 7) — additive helper, not in the binding list.
// ---------------------------------------------------------------------

#[derive(Debug, Clone, PartialEq)]
pub struct LegacyEnvelope {
    pub kind: String,
    pub legacy_kind: String,
    pub payload: Json,
    pub authenticated: bool,
    pub source_revision: Option<String>,
    pub source: String,
    pub captured_at: String,
    pub provisional_mapping_ref: Option<String>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ImportedLegacyEvidence {
    pub evidence_id: String,
    pub run_id: Option<String>,
    pub legacy_kind: String,
    pub capability: String,
    pub observation: Json,
    pub evidence_class: String,
    pub source_revision: Option<String>,
    pub authentication_issuer_identity: String,
    pub trust_class: String,
    pub never_qualifying: bool,
    pub provisional_mapping_ref: Option<String>,
    pub observed_at: String,
    pub stale: bool,
}

/// Import a legacy-envelope-v1-shaped record as an Arcane evidence envelope
/// that can never gain trust. Mirrors JS `importLegacyEvidence`.
///
/// `evidence_id` is caller-supplied (see module doc: minting is out of
/// this chunk's ownership) in place of `mintKernelId('evidenceReceipt')`.
pub fn import_legacy_evidence(
    legacy: &LegacyEnvelope,
    evidence_id: String,
    run_id: Option<String>,
    capability: Option<String>,
    observed_at: Option<String>,
) -> Result<ImportedLegacyEvidence, EnvelopeError> {
    if legacy.kind != "legion-legacy-envelope" {
        return Err(EnvelopeError {
            code: ArcCode::ArcSchemaInvalid,
            message: format!("expected a legacy-envelope-v1 shaped record (kind: legion-legacy-envelope), got: {}", legacy.kind),
        });
    }
    if legacy.authenticated {
        return Err(EnvelopeError {
            code: ArcCode::ArcAuthLegacyDigest,
            message: "legacy envelope provenance.authenticated must be false".to_string(),
        });
    }

    let payload_digest = digest_value(&legacy.payload);

    Ok(ImportedLegacyEvidence {
        evidence_id,
        run_id,
        legacy_kind: legacy.legacy_kind.clone(),
        capability: capability.unwrap_or_else(|| "legacy-import".to_string()),
        observation: Json::Obj(vec![("legacyPayloadDigest".into(), Json::str(payload_digest))]),
        evidence_class: "external".to_string(),
        source_revision: legacy.source_revision.clone(),
        authentication_issuer_identity: legacy.source.clone(),
        trust_class: "legacy-unauthenticated".to_string(),
        never_qualifying: true,
        provisional_mapping_ref: legacy.provisional_mapping_ref.clone(),
        observed_at: observed_at.unwrap_or_else(|| legacy.captured_at.clone()),
        stale: false,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    fn base_input() -> SealEvidenceInput {
        SealEvidenceInput {
            evidence_id: "ev-1".into(),
            provisional: false,
            run_id: "run-1".into(),
            task_id: None,
            contract_id: None,
            producer_authority: "capability:foo".into(),
            capability: "foo".into(),
            observation: Json::Obj(vec![("k".into(), Json::str("v"))]),
            evidence_class: "internal".into(),
            source_revision: Some("git:abc".into()),
            dependencies: vec![],
            authentication_verification_method: Some("mac".into()),
            authentication: Json::Null,
            replay_defense: Json::Null,
            observed_at: "2026-01-01T00:00:00Z".into(),
        }
    }

    #[test]
    fn seal_evidence_projects_dependencies_and_folds_dimension_into_ref() {
        let mut input = base_input();
        input.dependencies = vec![Dependency { dimension: "source-digest".into(), reference: "sha256:aa".into(), digest: Some("sha256:aa".into()) }];
        let sealed = seal_evidence(input).unwrap();
        let Json::Obj(receipt) = &sealed.receipt else { panic!() };
        let depends_on = receipt.iter().find(|(k, _)| k == "dependsOn").unwrap();
        let Json::Arr(items) = &depends_on.1 else { panic!() };
        assert_eq!(items.len(), 1);
        let Json::Obj(item) = &items[0] else { panic!() };
        assert_eq!(item.iter().find(|(k, _)| k == "kind").unwrap().1, Json::str("source-revision"));
        assert_eq!(item.iter().find(|(k, _)| k == "ref").unwrap().1, Json::str("source-digest:sha256:aa"));
    }

    #[test]
    fn seal_evidence_rejects_unknown_dimension() {
        let mut input = base_input();
        input.dependencies = vec![Dependency { dimension: "bogus".into(), reference: "x".into(), digest: None }];
        let err = seal_evidence(input).unwrap_err();
        assert_eq!(err.code, ArcCode::ArcDependencyUnknown);
    }

    #[test]
    fn seal_evidence_rejects_empty_dependency_ref() {
        let mut input = base_input();
        input.dependencies = vec![Dependency { dimension: "source-digest".into(), reference: "".into(), digest: None }];
        let err = seal_evidence(input).unwrap_err();
        assert_eq!(err.code, ArcCode::ArcSchemaInvalid);
    }

    #[test]
    fn seal_evidence_derives_unauthenticated_trust_class() {
        let mut input = base_input();
        input.authentication_verification_method = Some("unauthenticated".into());
        let sealed = seal_evidence(input).unwrap();
        let Json::Obj(env) = &sealed.envelope else { panic!() };
        assert_eq!(env.iter().find(|(k, _)| k == "trustClass").unwrap().1, Json::str("unauthenticated"));
    }

    #[test]
    fn envelope_digest_is_idempotent_across_pre_and_post_digested_envelope() {
        let sealed = seal_evidence(base_input()).unwrap();
        let redigested = envelope_digest(&sealed.envelope);
        assert_eq!(redigested, sealed.envelope_digest);
    }

    #[test]
    fn unfaithful_dimensions_matches_js_count() {
        // JS: 9 of 16 dimensions have no faithful bucket.
        assert_eq!(unfaithful_dimensions().len(), 9);
        assert_eq!(DEPENDENCY_DIMENSION.len(), 16);
    }

    #[test]
    fn import_legacy_evidence_rejects_wrong_kind() {
        let legacy = LegacyEnvelope {
            kind: "not-legacy".into(),
            legacy_kind: "x".into(),
            payload: Json::Null,
            authenticated: false,
            source_revision: None,
            source: "s".into(),
            captured_at: "2026-01-01T00:00:00Z".into(),
            provisional_mapping_ref: None,
        };
        let err = import_legacy_evidence(&legacy, "ev-2".into(), None, None, None).unwrap_err();
        assert_eq!(err.code, ArcCode::ArcSchemaInvalid);
    }

    #[test]
    fn import_legacy_evidence_rejects_authenticated_true() {
        let legacy = LegacyEnvelope {
            kind: "legion-legacy-envelope".into(),
            legacy_kind: "x".into(),
            payload: Json::Null,
            authenticated: true,
            source_revision: None,
            source: "s".into(),
            captured_at: "2026-01-01T00:00:00Z".into(),
            provisional_mapping_ref: None,
        };
        let err = import_legacy_evidence(&legacy, "ev-3".into(), None, None, None).unwrap_err();
        assert_eq!(err.code, ArcCode::ArcAuthLegacyDigest);
    }

    #[test]
    fn import_legacy_evidence_never_qualifies_and_is_unauthenticated() {
        let legacy = LegacyEnvelope {
            kind: "legion-legacy-envelope".into(),
            legacy_kind: "x".into(),
            payload: Json::str("p"),
            authenticated: false,
            source_revision: Some("git:1".into()),
            source: "legacy-tool".into(),
            captured_at: "2026-01-01T00:00:00Z".into(),
            provisional_mapping_ref: Some("map-1".into()),
        };
        let imported = import_legacy_evidence(&legacy, "ev-4".into(), None, None, None).unwrap();
        assert_eq!(imported.trust_class, "legacy-unauthenticated");
        assert!(imported.never_qualifying);
        assert_eq!(imported.observed_at, "2026-01-01T00:00:00Z");
    }
}
